rpc_new/provider/base.ts

```diff
@@ -0,0 +1,71 @@
+import * as msg from "../messages.ts";
+import * as U from "../util.ts";
+import { ProviderCloseError, ProviderHandlerError, ProviderSendError } from "./errors.ts";
+
+/**
+ * @param discoveryValue the value with which to discover the given chain
+ * @param listener the callback to which messages and errors should be applied
+ */
+export type Provider<DiscoveryValue, SendErrorData, HandlerErrorData, CloseErrorData> = (
+  discoveryValue: DiscoveryValue,
+  listener: ProviderListener<SendErrorData, HandlerErrorData>,
+) => ProviderRef<CloseErrorData>;
+
+export type ProviderListener<SendErrorData, HandlerErrorData> = U.Listener<
+  | msg.IngressMessage
+  | ProviderSendError<SendErrorData>
+  | ProviderHandlerError<HandlerErrorData>
+>;
+
+export interface ProviderRef<CloseErrorData> {
+  nextId(): string;
+  send(message: msg.EgressMessage): void;
+  release(): Promise<void | ProviderCloseError<CloseErrorData>>;
+}
+
+declare const providerListener_: unique symbol;
+type providerListener_ = typeof providerListener_;
+
+export abstract class ProviderConnection<Inner, SendErrorData, HandlerErrorData> {
+  declare [providerListener_]: ProviderListener<SendErrorData, HandlerErrorData>;
```

Comment:
This makes the class invariant on these type parameters

Reply:
What problems do you foresee this causing? It's just there for convenience. Can get rid if you think it'll be problematic.

Reply:
It's not *necessarily* a problem, but it is counter-semantic.